### PR TITLE
fix: Make `getActionFolderRole()` use the correct Folder in all cases

### PR DESCRIPTION
**Pull request by KevinBoulongne**
Depends on #2417

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/utils/FolderRoleUtils.kt [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    private fun getActionFolderRole(
+        folderId: String,
+        snoozable: Snoozable,
+        folderController: FolderController,
+    ): FolderRole? {
+        val folderRole = folderController.getFolder(folderId)?.role
+        return if (folderRole == FolderRole.INBOX && snoozable.isSnoozed()) FolderRole.SNOOZED else folderRole
+    }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/utils/FolderRoleUtils.kt]: I see you ended up changing your approach and you passed down the folderController through every method in the end instead of injecting the FolderRoleUtils through DI. What was the motivation behind this change?

File: app/src/main/java/com/infomaniak/mail/utils/FolderRoleUtils.kt
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/utils/FolderRoleUtils.kt:1]: We should definitely test the methods of this file because it's easier and more robust than trying to find a way to test everything inside of the app by hand in this case

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 2]
```diff
@@ -87,8 +88,8 @@ import com.infomaniak.lib.core.R as RCore
 class MainViewModel @Inject constructor(
     application: Application,
     avatarMergedContactData: AvatarMergedContactData,
+    val folderController: FolderController,
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt]: I wonder to what extent we want to expose a controller outside of a viewmodel and into views

File: app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/FolderController.kt [hunk 3]
```diff
@@ -44,6 +44,8 @@ class FolderController @Inject constructor(
     private val mailboxContentRealm: RealmDatabase.MailboxContent,
 ) {
 
+    var testRealm: Realm? = null
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/data/cache/mailboxContent/FolderController.kt]: We can check tomorrow but this should not be needed because you can pass the realm through the constructor in mailboxContentRealm, no? And I think that you already do it.

**Review by LunarX — CHANGES_REQUESTED** [review]
I only quickly glanced at it but I'll look into it tomorrow

File: app/src/main/java/com/infomaniak/mail/data/cache/RealmDatabase.kt [hunk 4]
```diff
@@ -110,13 +110,24 @@ object RealmDatabase {
...
+    // Used for tests only.
+    class TestMailboxContent : MailboxContent() {
+        override operator fun invoke() = runBlocking(Dispatchers.IO) {
+            mailboxContentMutex.withLock {
+                _mailboxContent ?: newMailboxContentInstance(
+                    userId = -42, mailboxId = -1337, loadDataInMemory = true,
+                ).also { _mailboxContent = it }
+            }
+        }
+    }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/data/cache/RealmDatabase.kt]: We can extract it outside of the "real files" and into the test package

**Review by LunarX — CHANGES_REQUESTED** [review]
first comment

File: app/src/main/java/com/infomaniak/mail/data/cache/RealmDatabase.kt [hunk 5]
```diff
@@ -110,13 +110,24 @@ object RealmDatabase {
...
         }
     }
+
+    // Used for tests only.
+    class TestMailboxContent : MailboxContent() {
+        override operator fun invoke() = runBlocking(Dispatchers.IO) {
+            mailboxContentMutex.withLock {
+                _mailboxContent ?: newMailboxContentInstance(
+                    userId = -42, mailboxId = -1337, loadDataInMemory = true,
+                ).also { _mailboxContent = it }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/data/cache/RealmDatabase.kt]: We should be able to call `RealmConfig.mailboxContent` directly and leave `newMailboxContentInstance` intact

File: app/src/main/java/com/infomaniak/mail/data/models/Folder.kt [hunk 6]
```diff
@@ -56,7 +56,7 @@ class Folder : RealmObject, Cloneable {
     var path: String = ""
     var name: String = ""
     @SerialName("role")
-    private var _role: String? = null
+    var _role: String? = null // Don't edit this value, it's only public for tests.
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/data/models/Folder.kt]: We could create a "annotation" package with the following annotation to use here

```kt
@RequiresOptIn(
    level = RequiresOptIn.Level.ERROR,
    message = "Don't access it outside of testing"
)
@Retention(AnnotationRetention.BINARY)
@Target(AnnotationTarget.PROPERTY)
annotation class TestOnly
```

File: app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+import org.junit.runner.RunWith
+import org.mockito.Mock
+import org.mockito.Mockito.mock
+import org.mockito.junit.MockitoJUnitRunner
+
+@RunWith(MockitoJUnitRunner::class)
+class FolderRoleUtilsTest {
+
+    @Mock
+    private val mockContext = mock<Context>()
```
Comment (LunarX) [inline app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt]: Also we should add a comment explaining that we don't need the context because it's only used in methods that we don't call during this test

File: app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    private val mailboxContentRealm = DummyMailboxContent()
+    private val folderController = FolderController(mockContext, mailboxContentRealm)
+    private val folderRoleUtils = FolderRoleUtils(folderController)
+
+    private fun setup() {
+        mailboxContentRealm().writeBlocking {
+            copyToRealm(folderInbox, UpdatePolicy.ALL)
+            copyToRealm(folderDraft, UpdatePolicy.ALL)
+        }
+    }
```
Comment (LunarX) [inline app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt]: There's usually a way for test to run a specific bit of code before each test, can we use it here?

File: app/src/test/java/com/infomaniak/mail/dataset/DummyThreads.kt [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import com.infomaniak.mail.dataset.DummyMessages.messageDraft
+import com.infomaniak.mail.dataset.DummyMessages.messageInbox
+import com.infomaniak.mail.dataset.DummyMessages.messageInboxSnoozed
+
+object DummyThreads {
+
+    private val mailboxContentRealm = DummyMailboxContent()
+
+    val threadInboxSnoozed = threadOf(messageInbox, messageInboxSnoozed)
+    val threadDraft = threadOf(messageDraft)
```
Comment (LunarX) [inline app/src/test/java/com/infomaniak/mail/dataset/DummyThreads.kt]: Maybe we should also test how this works when dealing with "search" threads and their different edge-cases like a single message of any folder vs a single message of the snooze folder or a single message and the case were we can get threads of multiple messages at once

File: app/src/test/java/com/infomaniak/mail/dataset/DummyFolders.kt [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+object DummyFolders {
+
+    const val FOLDER_INBOX_ID = "FOLDER_INBOX_ID"
+    const val FOLDER_DRAFT_ID = "FOLDER_DRAFT_ID"
+
+    val folderInbox = Folder().apply {
+        id = FOLDER_INBOX_ID
+        _role = FolderRole.INBOX.name
+
+        threadInboxSnoozed.folderId = id
```
Comment (LunarX) [inline app/src/test/java/com/infomaniak/mail/dataset/DummyFolders.kt]: Does it make sense to set this thread property inside of the folder initialization?

File: app/src/test/java/com/infomaniak/mail/dataset/DummyThreads.kt [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+package com.infomaniak.mail.dataset
+
+import com.infomaniak.mail.data.models.message.Message
+import com.infomaniak.mail.data.models.thread.Thread
+import com.infomaniak.mail.dataset.DummyMessages.messageDraft
+import com.infomaniak.mail.dataset.DummyMessages.messageInbox
+import com.infomaniak.mail.dataset.DummyMessages.messageInboxSnoozed
+
+object DummyThreads {
+    val threadInboxSnoozed = threadOf(messageInbox, messageInboxSnoozed)
```
Comment (LunarX) [inline app/src/test/java/com/infomaniak/mail/dataset/DummyThreads.kt]: This `threadOf(messageInbox, messageInboxSnoozed)` situation is actually impossible in practice. A thread in inbox will always have either all of its inbox message be snoozed or all of them be not snoozed.

Non-snoozed messages in a snoozed thread will always be from another folder

File: app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt (hunk 8 above)
Comment (LunarX) [reply app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt]: Also, if we need a setup then don't we need an associated reciprocate tear down?

File: app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+        assertTrue(folderRole == FolderRole.DRAFT)
+    }
+
+    @Test
+    fun threadInboxSnoozed_should_be_SNOOZED() {
+        setup()
+
+        val folderRole = folderRoleUtils.getActionFolderRole(threadInboxSnoozed)
+        assertTrue(folderRole == FolderRole.SNOOZED)
+    }
```
Comment (LunarX) [inline app/src/test/java/com/infomaniak/mail/FolderRoleUtilsTest.kt]: We need a test of a snoozed thread that's inside of the snooze folder to correctly return that it's folder role is snooze

File: app/src/test/java/com/infomaniak/mail/dataset/DummyFolders.kt (hunk 10 above)
Comment (LunarX) [inline app/src/test/java/com/infomaniak/mail/dataset/DummyFolders.kt]: A snoozed thread cannot be inside the inbox folder, it will be inside the snooze folder

**Review by LunarX — APPROVED** [review]
